Queue Splitting Feature

t4lz opened the pull request:
#2066

aviramha commented on mirrord/config/src/feature/split_queues.rs:
```diff
@@ -0,0 +1,71 @@
+/// Set of rules to filter that are required to apply to a single field
+enum SQSAttributeRuleSet {
+    /// A single regex is enough for complex rules for a string.
+    StringAttribute(String),
```
I think we can start with String attribute only to make it a bit easier

aviramha commented on mirrord/config/src/feature/split_queues.rs:
```diff
@@ -0,0 +1,71 @@
+/// {
+///   "feature": {
+///     "split_queues": {
+///       "whatever-q-splitter": {
```
what's this?

aviramha commented on mirrord/config/src/feature/split_queues.rs:
```diff
@@ -0,0 +1,71 @@
+use mirrord_analytics::CollectAnalytics;
+use mirrord_config_derive::MirrordConfig;
+
+/// Maps name of a `MirrordQueueSplitter` resource to a mapping from a fork name to a filter
```
fork -> splitter

aviramha commented on mirrord/operator/src/crd.rs:
```diff
@@ -189,3 +191,71 @@ pub struct MirrordPolicySpec {
+    #[serde(rename_all = "camelCase")] // config_map -> configMap in yaml.
+    SQS {
+        /// AWS region to use.
+        region: String,
```
I think we just use the region we're in?

aviramha replied:
like I'm pretty sure it's part of the credentials provided to us

Razz4780 commented on mirrord/config/src/feature/split_queues.rs:
```diff
@@ -0,0 +1,85 @@
+#[config(derive = "JsonSchema,Eq,PartialEq")]
+pub struct SplitQueuesConfig {
+    /// The name of the `MirrordQueueSplitter` resource to get queue details from.
+    splitter: String,
```
Not sure yet about the architecture - will the cluster always have one splitter? If not, does it make sense for the user app to use more than one splitter?

t4lz replied:
There can be one splitter per deployment (or Argo rollout). The app won't have to specify the splitter, it will be automatically derived from the target. I still haven't updated the code since the meeting where that was decided.

Razz4780 commented on mirrord/config/src/feature/split_queues.rs:
```diff
@@ -0,0 +1,85 @@
+#[derive(Serialize, Deserialize, JsonSchema, Clone, Debug, PartialEq, Eq)]
+#[serde(tag = "queue_type")]
+enum QueueMessageFilter {
+    SQS(HashMap<String, SQSAttributeRuleSet>),
```
This needs a doc, I think. To let the user know what this mapping actually means (I don't know what SQS attribute is, probably some devs don't either).

Razz4780 commented on mirrord/config/src/feature/split_queues.rs:
```diff
@@ -0,0 +1,85 @@
+    // GreaterThanFloat(f64),
+    // LesserThanFloat(f64),
+    EqualsInt(i64),
+    EqualsFloat(i64),
```
```suggestion
    EqualsFloat(f64),
```
Right?

t4lz replied:
Yes, thanks. Though typed rules will probably be dropped from this PR, following https://github.com/metalbear-co/mirrord/pull/2173#discussion_r1451977234.

Razz4780 commented on mirrord/config/src/feature/split_queues.rs:
```diff
@@ -0,0 +1,85 @@
+}
+
+#[derive(Serialize, Deserialize, JsonSchema, Clone, Debug, PartialEq, Eq)]
+enum BinaryAttributeRule {
```
I'd add a doc to explain encoding of these rules, because we're matching against binary data.

Razz4780 commented on mirrord/config/src/feature/split_queues.rs:
```diff
@@ -0,0 +1,85 @@
+
+#[derive(Serialize, Deserialize, JsonSchema, Clone, Debug, PartialEq, Eq)]
+/// Set of rules to filter that are required to apply to a single field
+enum SQSAttributeRuleSet {
```
```suggestion
#[serde(tag = "type", content = "value", rename_all = "snake_case")
enum SQSAttributeRuleSet {
```
Suggestion

Razz4780 commented on mirrord/config/src/feature/split_queues.rs:
```diff
@@ -0,0 +1,85 @@
+/// Set of rules to filter that are required to apply to a single field
+enum SQSAttributeRuleSet {
+    /// A single regex is enough for complex rules for a string.
+    StringAttribute(String),
```
```suggestion
    MatchesRegex(String),
```

Razz4780 commented on mirrord/operator/src/client.rs:
```diff
@@ -642,16 +646,26 @@ impl OperatorApi<PreparedClientCert> {
+        self.check_feature_support(config)?;
 
-        let target = if config.feature.copy_target.enabled {
+        let target = if config.feature.copy_target.enabled
+            // use copy_target for splitting queues
```
I don't think this should be done transparently. I think if `copy_target` is not enabled it's a config verification error. @aviramha wdyt?

Razz4780 replied:
Wdyt about a warning then?

Razz4780 replied:
Ok, let it be `tracing::info!`

t4lz replied:
added.

Razz4780 commented on mirrord/config/src/feature/split_queues.rs:
```diff
@@ -0,0 +1,96 @@
+
+    /// Out of the whole queue splitting config, get only the sqs queues.
+    pub fn get_sqs_filter(&self) -> Option<HashMap<String, SqsMessageFilter>> {
+        self.0.as_ref().map(|queue_id2queue_filter| {
```
I don't know if I like the name of this lambda argument :V

Razz4780 commented on mirrord/config/src/feature/split_queues.rs:
```diff
@@ -0,0 +1,96 @@
+                    QueueFilter::Sqs(filter_mapping) => {
+                        Some((queue_id.clone(), filter_mapping.clone()))
+                    }
+                    #[allow(unreachable_patterns)] // will become reachable when we add variants
+                    _ => None,
```
Maybe just remove this branch?

Razz4780 commented on mirrord/config/src/feature/split_queues.rs:
```diff
@@ -0,0 +1,96 @@
+/// {
+///   "feature": {
+///     "split_queues": {
+///       "first-queue": {
+///         "queue_type": "SQS",
+///         "message_filter": {
+///           "wows": "so wows",
+///           "coolz": "^very .*"
+///         }
+///       },
```
```suggestion
///       "SQS/first-queue": {
///         "message_filter": {
///           "wows": "so wows",
///           "coolz": "^very .*"
///         }
///       },
```
I think following `target` pattern would make this more concise

Razz4780 replied:
Now that I look at it again, I think it would be too much of a hassle.
`QueueId` would have to store queue type and then we'd have to manually check that the value variant matches. `untagged` sounds a bit scary, because we'd have to ensure that no two variants can be parsed from the same json.
tl;dr it's ok as it is

Razz4780 commented on mirrord/config/src/feature/split_queues.rs:
```diff
@@ -0,0 +1,96 @@
+#[non_exhaustive]
+pub enum QueueFilter {
+    #[serde(rename = "SQS")]
+    Sqs(SqsMessageFilter),
```
We should document whether this set of filters means a conjunction or an alternative.
Also, it could use even a short description, like "Amazon Simple Queue Service"

Razz4780 commented on mirrord/operator/src/crd.rs:
```diff
@@ -137,9 +238,48 @@ pub struct Session {
+    /// This variant is what a client sees when the operator includes a feature the client is not
+    /// yet aware of, because it was introduced in a version newer than the client's.
+    #[serde(other)]
+    FeatureFromTheFuture,
```
```suggestion
    Unknown,
```

t4lz replied:
Changed + added "No time travel" to style guide.

Razz4780 commented on mirrord/operator/src/crd.rs:
```diff
@@ -137,9 +238,48 @@ pub struct Session {
+pub enum NewOperatorFeature {
+    ProxyApi,
+    CopyTarget,
+    Sqs,
```
```suggestion
    SqsQueueSplitting,
```

Razz4780 commented on mirrord/operator/src/crd.rs:
```diff
@@ -161,6 +301,8 @@ pub struct CopyTargetSpec {
     /// Ignored if [`Target`] is not [`Target::Deployment`].
     pub scale_down: bool,
+    /// queue id -> (attribute name -> regex)
+    pub sqs_filter: Option<HashMap<QueueId, SqsMessageFilter>>,
```
Maybe we should use a nested struct here? To group all queue filters in a separate type

t4lz replied:
Should I just send over the whole `SplitQueuesConfig`, and do the partitioning by queue types in the operator's side?

Razz4780 replied:
Yeah, sounds good

Razz4780 commented on mirrord/operator/src/crd.rs:
```diff
@@ -199,3 +341,190 @@ pub struct MirrordPolicySpec {
+    EnvVar(String),
+}
+
+pub type OutputQueueName = String;
```
Looks like this is not used

Razz4780 commented on mirrord/operator/src/crd.rs:
```diff
@@ -199,3 +341,190 @@ pub struct MirrordPolicySpec {
+/// The details of a queue that should be split.
+#[derive(Clone, Debug, Deserialize, Eq, PartialEq, Serialize, JsonSchema)]
+#[serde(tag = "queueType")]
+// So that controllers that only handle 1 type of queue don't have to adapt when we add more queue
+// types.
+#[non_exhaustive]
```
I'm not a big fan of this, but I won't complain :v

t4lz replied:
What's the problem with that?

Razz4780 commented on mirrord/operator/src/crd.rs:
```diff
@@ -199,3 +341,190 @@ pub struct MirrordPolicySpec {
+/// A workload that is a consumer of a queue that is being split.
+#[derive(Clone, Debug, Deserialize, Eq, PartialEq, Serialize, JsonSchema)]
+#[serde(rename_all = "camelCase")] // Deployment -> deployment in yaml.
+pub enum QueueConsumer {
+    Deployment(String),
+    Rollout(String),
+}
```
What about jobs/cronjobs/statefulsets? We're able to patch those as well

t4lz replied:
Let's add support for those in a separate step.
Is it ok to change the CRD later, or do I have to add them here even if we don't support them yet (and return an error if users try to use them)?

Razz4780 replied:
I think it's ok to change them later

Razz4780 commented on mirrord/operator/src/crd.rs:
```diff
@@ -199,3 +341,190 @@ pub struct MirrordPolicySpec {
+    // it here in the form that is useful to reader, for simplicity and readability.
+    /// Env var name -> old and new queue names.
+    pub env_updates: BTreeMap<String, QueueNameUpdate>,
+    // TODO: add updates to config map references.
```
todo :V

t4lz replied:
We're going to add support for config map queue name sources in a separate step, this is a deliberate placeholder for when we do that, can I leave it there or should I remove it?

Razz4780 commented on mirrord/operator/src/crd.rs:
```diff
@@ -199,3 +341,190 @@ pub struct MirrordPolicySpec {
+    // TODO: add updates to config map references.
+}
+
+/// The [`kube::runtime::wait::Condition`] trait is auto-implemented for this function.
+/// To be used in [`kube::runtime::wait::await_condition`].
+pub fn is_session_ready(session: Option<&MirrordSqsSession>) -> bool {
+    session
+        .map(|session| session.status.is_some())
+        .unwrap_or_default()
+}
```
I don't think this should be here, operator code seems like a better fit

t4lz replied:
Really? But it depends on the definition of the status struct, and might have to be changed if we change that struct. I think having it here makes more sense and could also help people remember to change it if they change the struct.

Razz4780 reviewed (changes requested):
I think some updates are needed in `LayerConfig::verify_config` command - queue splitting is not compatible with all target types and require operator.

Also, I'd make sure that all crds are forward compatible, as we're gonna add more queue types. Old CLIs/operators must not fail to deserialize crds after update

t4lz commented on mirrord/operator/src/setup.rs:
```diff
@@ -432,6 +438,20 @@ impl OperatorRole {
                     verbs: vec!["get".to_owned(), "list".to_owned(), "watch".to_owned()],
                     ..Default::default()
                 },
+                // For SQS controller to temporarily change deployments to use changed queues.
+                PolicyRule {
+                    api_groups: Some(vec!["apps".to_owned()]),
+                    resources: Some(vec!["deployments".to_owned()]),
+                    verbs: vec!["patch".to_owned()],
+                    ..Default::default()
+                },
+                // For SQS controller to temporarily change Argo Rollouts to use changed queues.
+                PolicyRule {
+                    api_groups: Some(vec!["argoproj.io".to_owned()]),
+                    resources: Some(vec!["rollouts".to_owned()]),
+                    verbs: vec!["patch".to_owned()],
+                    ..Default::default()
+                },
```
Let's do that in a separate PR.

Razz4780 commented on mirrord/config/src/feature/split_queues.rs:
```diff
@@ -0,0 +1,97 @@
+pub type AttributeValuePattern = String;
+
+/// A filter is a mapping between message attribute names and regexes they should match.
+/// The local application will only receive messages that match ALL the patterns.
```
```suggestion
/// The local application will only receive messages that match **all** of the given patterns.
```

Razz4780 commented on mirrord/config/src/feature/split_queues.rs:
```diff
@@ -0,0 +1,97 @@
+
+/// A filter is a mapping between message attribute names and regexes they should match.
+/// The local application will only receive messages that match ALL the patterns.
+/// This means, only messages that have all the `MessageAttributeName`s in the filter,
```
```suggestion
/// This means, only messages that have **all** the `MessageAttributeName`s in the filter,
```

Razz4780 commented on mirrord/operator/src/crd.rs:
```diff
@@ -199,3 +342,206 @@ pub struct MirrordPolicySpec {
+#[derive(Clone, Debug, Deserialize, Eq, PartialEq, Serialize, JsonSchema)]
+#[serde(rename_all = "camelCase")] // Deployment -> deployment in yaml.
+pub enum QueueConsumer {
+    Deployment(String, Option<String>),
+    Rollout(String, Option<String>),
```
What does it mean when the container is not given? Needs a doc I think

t4lz commented on mirrord/operator/src/crd.rs:
```diff
@@ -199,3 +341,188 @@ pub struct MirrordPolicySpec {
+    /// in this MirrordWorkloadQueueRegistry, alongside with the original tags of the respective
+    /// original queue. In case of a collision, the temporary queue will get the value from the
+    /// tag passed in here.
+    pub tags: Option<HashMap<String, String>>,
```
Moved, also changed operator side accordingly

t4lz commented on mirrord/operator/src/crd.rs:
```diff
@@ -199,3 +341,188 @@ pub struct MirrordPolicySpec {
+    /// The id of the mirrord exec session, from the operator.
+    // The Kubernetes API can't deal with 64 bit numbers (with most significant bit set)
+    // so we save that field as a string even though its source is a u64
+    pub session_id: String,
```
done (in operator side)

t4lz replied:
Changed to lower hex now, because I'm using it in resource names (of `MirrordSqsSession`) and resource names have to be lower case.

Razz4780 reviewed (approved):
Needs rebasing ^^

t4lz commented:
What changed besides the requested changes:
- Added `#[schemars(skip)]` `#[serde(skip_serializing)]` to unknown (`#[serde(other)]`) CRD enum variants, so that they are not part of the schema, but only deserialized when a peer has a newer version and sends a new, unknown variant.
- Changed `QueueConsumer` to a struct because that's now a better representation after having added a container.

Razz4780 reviewed (changes requested):
Only some small ser/de changes

t4lz commented on mirrord/operator/src/crd.rs:
```diff
@@ -199,3 +342,271 @@ pub struct MirrordPolicySpec {
+
+/// A workload that is a consumer of a queue that is being split.
+#[derive(Clone, Debug, Deserialize, Eq, PartialEq, Serialize, JsonSchema)]
+pub enum QueueConsumerType {
```
Why snake case? These variants are object types, which I think are UpperCamelCase in k8s.

![image](https://github.com/user-attachments/assets/dd8725d8-14da-42c4-ac25-79e44f15233c)
https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#types-kinds

Razz4780 replied:
Ah, right. Sorry
